File: Packages/OpenVibbleKit/Sources/BuddyStats/PersonaStats.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation

public struct PersonaStats: Codable, Sendable, Equatable {
    public static let tokensPerLevel: UInt32 = 50_000
    public static let velocityRingSize: Int = 8

    public var approvals: UInt16
    public var denials: UInt16
    public var velocity: [UInt16]
    public var velIdx: UInt8
    public var velCount: UInt8
    public var level: UInt8
    public var tokens: UInt32
    public var napSeconds: UInt32

    public init(
        approvals: UInt16 = 0,
        denials: UInt16 = 0,
        velocity: [UInt16] = Array(repeating: 0, count: 8),
        velIdx: UInt8 = 0,
        velCount: UInt8 = 0,
        level: UInt8 = 0,
        tokens: UInt32 = 0,
        napSeconds: UInt32 = 0
    ) {
        self.approvals = approvals
        self.denials = denials
        self.velocity = velocity
        self.velIdx = velIdx
        self.velCount = velCount
        self.level = level
        self.tokens = tokens
        self.napSeconds = napSeconds
    }

    public var derivedLevel: UInt8 {
        UInt8(min(UInt32(UInt8.max), tokens / Self.tokensPerLevel))
    }

    public var fedProgress: UInt8 {
        // 0..9: tokens within current level / (level_size/10)
        let partial = tokens % Self.tokensPerLevel
        let perPip = Self.tokensPerLevel / 10
        return UInt8(min(9, partial / perPip))
    }

    public var medianVelocitySeconds: UInt16 {
        guard velCount > 0 else { return 0 }
        let slice = velocity.prefix(Int(velCount))
        let sorted = slice.sorted()
        return sorted[Int(velCount) / 2]
    }

    /// 0..4 tier. Faster median response = higher; heavy deny rate pulls it down.
    public var moodTier: UInt8 {
        var tier: Int
        let vel = medianVelocitySeconds
        if vel == 0 { tier = 2 }
        else if vel < 15 { tier = 4 }
        else if vel < 30 { tier = 3 }
        else if vel < 60 { tier = 2 }
        else if vel < 120 { tier = 1 }
        else { tier = 0 }
        let a = Int(approvals), d = Int(denials)
        if a + d >= 3 {
            if d > a { tier -= 2 }
            else if d * 2 > a { tier -= 1 }
        }
        return UInt8(max(0, min(4, tier)))
    }
}

File: android/buddy-stats/src/main/kotlin/com/openvibble/stats/PersonaStats.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.openvibble.stats

import kotlinx.serialization.Serializable

/**
 * Stored persona state. iOS uses fixed-width ints (UInt16/UInt8/UInt32)
 * that map into a single persisted 14-byte record; here we persist via
 * DataStore so plain Int/Long is fine, but we still clamp to the same
 * ranges to keep the derived stats math identical across platforms.
 */
@Serializable
data class PersonaStats(
    val approvals: Int = 0,
    val denials: Int = 0,
    val velocity: List<Int> = List(VELOCITY_RING_SIZE) { 0 },
    val velIdx: Int = 0,
    val velCount: Int = 0,
    val level: Int = 0,
    val tokens: Long = 0L,
    val napSeconds: Long = 0L,
) {
    val derivedLevel: Int
        get() = (tokens / TOKENS_PER_LEVEL).coerceIn(0, UByte.MAX_VALUE.toLong()).toInt()

    /** 0..9 level progress pip. */
    val fedProgress: Int
        get() {
            val partial = tokens % TOKENS_PER_LEVEL
            val perPip = TOKENS_PER_LEVEL / 10
            return (partial / perPip).coerceAtMost(9L).toInt()
        }

    val medianVelocitySeconds: Int
        get() {
            if (velCount <= 0) return 0
            val slice = velocity.take(velCount).sorted()
            return slice[velCount / 2]
        }

    /**
     * 0..4 tier: faster median response = higher, heavy deny rate pulls down.
     * Matches iOS `moodTier` 1:1 including the minimum-sample threshold.
     */
    val moodTier: Int
        get() {
            val vel = medianVelocitySeconds
            var tier = when {
                vel == 0 -> 2
                vel < 15 -> 4
                vel < 30 -> 3
                vel < 60 -> 2
                vel < 120 -> 1
                else -> 0
            }
            if (approvals + denials >= 3) {
                when {
                    denials > approvals -> tier -= 2
                    denials * 2 > approvals -> tier -= 1
                }
            }
            return tier.coerceIn(0, 4)
        }

    companion object {
        const val TOKENS_PER_LEVEL: Long = 50_000L
        const val VELOCITY_RING_SIZE: Int = 8
    }
}
